Create a signal to unassign roles from user when the role is deleted

D0rs4n opened the pull request:
Add a signal to the api app that automatically unassigns
deleted roles from users that have them
Closes #369

D0rs4n committed:
Create a signal to unassign roles from user when deleted

Add a signal to the api app that automatically unassigns
deleted roles from users that have them

Bluenix2 commented on pydis_site/apps/api/signals.py:
```diff
@@ -0,0 +1,12 @@
+@receiver(signal=pre_delete, sender=Role)
+def delete_role_from_user(sender: Role, instance: Role, **kwargs) -> None:
+    """Unassigns the Role (instance) that is being deleted from every user that has it."""
+    for user in User.objects.filter(roles__contains=[instance.id]):
+        del user.roles[user.roles.index(instance.id)]
+        user.save()
```
Is there not a way to do this in batch? We should let SQL do this

D0rs4n replied:
Oh, I see your concern. That wasn't a wise choice indeed. Although, I think there's a way to solve this using Django ORM.

D0rs4n replied:
It seems like, the Django ORM, F objects in particular does not support ArrayFields. Hence that's the only way to solve it via the Django ORM, for now.
However, I've never really worked with arrayfields using raw SQL.

jchristgit replied:
This isn't possible with plain PostgreSQL, except if we were to use triggers, and I would currently stay away from that personally.

The original approach here was the correct way: foreign key constraints. But since this got removed due to performance issues (waiting on reply from a Django expert friend on this topic), this is a fine approach.

D0rs4n committed:
Add test to check role unassignment

Create a test that checks if a role gets deleted it will also get unassigned from the user

D0rs4n committed:
Patch roles test to use fresh instance from the DB

D0rs4n committed:
Patch signals to use post_delete, instead of pre_delete

From now on the signal will only get executed after the Role has been deleted
The commit also introduces minor changes in the tests of roles

jchristgit reviewed (approved):
This looks good, thank you!

Could you rebase off master, please?

ChrisLovering reviewed (approved):
The seems to work well locally 👌
